Add documentation for form limits & improve configuration via context attributes (#12560)
eclipse/jetty.project: documentation/jetty/modules/programming-guide/pages/security/configuring-form-size.adoc

//
// ========================================================================
// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
// ========================================================================
//

> [[configuring-form-size]]

Reviewer: I'd rename it to `limit-form-content`, like the title.

= Limiting Form Content

Form content sent to the server is processed by Jetty into a map of parameters to be used by the web application.

> This can be vulnerable to denial of service (DOS) attacks since significant memory and CPU can be consumed if a malicious clients sends very large form content or large number of form keys.

Reviewer: ```suggestion
Forms can be a vector for denial-of-service attacks, since significant memory and CPU can be consumed if a malicious clients sends very large form content or a large number of form keys.
```

Thus, Jetty limits the amount of data and keys that can be in a form posted to Jetty.

The default maximum size Jetty permits is 200000 bytes and 1000 keys.

> You can change this default for a particular webapp or for all webapps on a particular Server instance.

Reviewer: ```suggestion
You can change this default for a particular web application or for all web applications on a particular `Server` instance.
```

> == Configuring Form Limits for a Webapp

Reviewer: ```suggestion
== Configuring Form Limits for a Web Application
```

> To configure the form limits for a single web application, the servlet context handler (or webappContext) instance must be configured using the following methods:

Reviewer: ```suggestion
To configure the form limits for a single web application, the `ServletContextHandler` (or `WebappContext`) instance must be configured using the following methods:
```

[,java,indent=0]
----
include::code:example$src/main/java/org/eclipse/jetty/docs/programming/security/FormSizeDocs.java[tags=formSizeConfig]
----

These methods may be called directly when embedding Jetty, but more commonly are configured from a context XML file or WEB-INF/jetty-web.xml file:

[,xml,subs=attributes+]
----
<Configure class="org.eclipse.jetty.{ee-current}.webapp.WebAppContext">

  ...

  <Set name="maxFormContentSize">200000</Set>
  <Set name="maxFormKeys">200</Set>
</Configure>

> ----

Reviewer: This should be moved to the operations guide.

> These settings can also be set via the following Context attributes.

Reviewer: It is not clear what "Context" are you referring to here.

You mean context attributes in `web.xml`?
If so, better have an example, but should propably better be in the operations guide.

> - `org.eclipse.jetty.server.Request.maxFormContentSize`.

Reviewer 2: Change as I did above: say what the config is and then say what the default is (no "however").
But why are we saying the same thing twice?   Sure one document can refer to the other?

Author: I removed duplicated sections and linked to programming guide

- `org.eclipse.jetty.server.Request.maxFormKeys`
- `org.eclipse.jetty.server.Request.maxFormContentSize`

== Configuring Default Form Limits for the Server

The default `maxFormKeys` is 1000 and the default `maxFormContentSize` is 200000.

> However, the following system properties can be set to change the default values of this across every context; `org.eclipse.jetty.server.Request.maxFormKeys` and `org.eclipse.jetty.server.Request.maxFormContentSize`.

Reviewer: Put them in a bullet list like above.

I wonder why we don't look these up as `Server` attributes, before using system properties?

> These methods take parameters for `maxFields` and `maxLength` which can be used to limit the form content.

Reviewer 2: They don't always take those parameters.  There are versions that use the defaults and they should be explained first and then say that the limits can be set problematically with this API

Author: Explained with comments in the code sample.
